content/blog/2019-10-02-probabril/index.md

```diff
@@ -0,0 +1,325 @@
+
+#### How is this possible given Abel-Ruffini?
+
+While probabilistic bril programs are not deterministic, they are far from being arbitrary matrices --- because they're probablistic transition matrices $A$ must have $\sum_{i} A_{i,j} = 1$ for all $j \in \mathcal S$. Moreover, because we already know the eigenvalues and even most of the eigenvectors we care about, there is no need to do some of this computation. 
+
+
+## Implementation
+
+This algorithm has been implemented in a file called `xbrili`, next to the normal interpreter `brili`. There are several additional features of note in the implemented version beyond what has already been discussed. First, the more major ellided details and extensions:
+
+- I have also implemented an extension of this algorithm which backs off of finite space assumptions, by specifying a tolerance $t$. When the mass of a state $u$ from the state state $s$ drops off enough, $\Pr(u \mid \vec{s}) < t$, we stop keeping track of the run entirely. It was actually rather hard to get this part right, because 
+- I have also implemented a utility called `random-bril.py` which generates random bril programs, with a mechanism for making sure that the distribution of different kinds of instructions is fairly even. There are a lot of parameters, so they are hard-coded. The parameters have been tuned to make things that look like reasonable programs to me, but of course do not reflect any real distribution of instructions, and there may be programs which are impossible to generate with it. Also, due to jumps, it is possible to generate programs that refer to variables before they are defined. We throw out any programs that generate errors like this.
+
+And also some more minor or more technical details that may be of interest:
+ 
+ - `brili` now interprets `flip` as random, `obv` as a restart, and `xbrili` introduces new actions for world splitting. 
+ - printing is now delayed until the end of the execution and saved in the buffer, so that observations truly can totally undo everything and allow for a restart.
+ - there are options for removing the printing instructions entirely, and dumping environment variables at the end, for both interpreters, so that we can automatically test the correctness of the two against one another via `random-bril`.
+ - I have also added quite a bit to `util.ts`, including a way of indexing typescript Maps by other typescript Maps by means of stringification.
+ 
+ 
+
+## Difficulties
+
+This took me a long time, and I was trying to figure out whether the algorithm I had in mind was sound, and how to get it to terminate, particularly once I didn't want to assume that my state space as finite right at the beginning. 
```

Comment:
Also, while this summary is useful to have and perhaps authentically describes the primary difficulty, I included this question in the guidelines while hoping for a few more specifics. It's intended to get you to talk in detail about shortcomings you see in the ideas and implementation. No need to expand this now, but I hope you'll be a little more specific in future iterations of this "difficulties" discussion.

Reply:
I maintain that my original proposal was not out of scope. I chose something that I mostly knew how to do, and then I was unable to avoid doing the exact inference for loops side quest which became the primary deliverable of the project.

I have expanded a small amount on this section